src/archivematicaCommon/lib/elasticSearchFunctions.py

```diff
@@ -635,8 +658,10 @@ def _generator():
     # It should be revisited once we make documents smaller.
     bulk(client, _generator(), chunk_size=50)
 
+    return (len(files), accession_ids)
 
-def index_transfer_and_files(client, uuid, path, printfn=print):
+
+def index_transfer_and_files(client, uuid, path, pending_deletion=False, printfn=print):
```

Comment:
There was some of the backlog work here, as I based my work on that since the backlog PR hadn't been merged anywhere when I started and there were some shared dependencies. My apologies if some of it remained! I'll include auditing for remaining bits from the backlog PR as part of my revisions.

You raise a good point about the inconsistency! My thinking was that AIP indexing should look up accession numbers in the METS XML for two reasons:

1. The AIP indexing function is already doing quite a lot of METS processing, including getting other information from the file `amdSec`s, so the additional lookup time for getting the accession number is pretty minimal (database lookups would almost certainly be quicker, but rewriting all of the AIP indexing seems out of scope for this project).

2. Since we support [indexing from the AIPs directly](https://github.com/artefactual/archivematica/blob/0722067836c3430f2a371f15583be633a6db7fe5/src/dashboard/src/main/management/commands/rebuild_elasticsearch_aip_index_from_files.py) even when there's no record of the AIP in the database, it seemed like the better approach.

I notice now that the reindexing script contains [some logic](https://github.com/artefactual/archivematica/blob/0722067836c3430f2a371f15583be633a6db7fe5/src/dashboard/src/main/management/commands/rebuild_transfer_backlog.py#L306) which handles this issue/edge case in a different way, by writing the information such as the accession number from the METS into the database prior to indexing if there's not already a database entry for the transfer. That could potentially be an interesting tack to consider taking with the AIPs as well if we were to rewrite the indexing functions?

Reply:
> There was some of the backlog work here, as I based my work on that since the backlog PR hadn't been merged anywhere when I started and there were some shared dependencies. My apologies if some of it remained! I'll include auditing for remaining bits from the backlog PR as part of my revisions.

I see. This is really a Git hygiene thig. I think it's definitely complicated here by tackling these two features, independently, but also sharing dependencies. Ideally you want to be working off a clean git reference, so I think the backlog's last commit ref is https://github.com/artefactual/archivematica/commit/ef6931b7a789b6009867bb607c58495cde246e98 so at some point you need to rebase against that if you're working against it and features of it are part of the PR to review.  

There becomes a clean delineation between concerns. 

The biggest impact is really the potential to introduce changes to the previous work (re-write, add new lines/ formatting, etc.). But I think we should find that eventual consistency when rebasing here. So rebase might be another next step. I don't think you'll see merge conflicts, but I think that can be the other risk of manually bringing consistency. 

It's a bit of a dance, and another point we'll need to consider if our feature merge strategies are modified. 

This is what I see in my head about how we've asked this work be approached:

![image](https://user-images.githubusercontent.com/1880412/82568216-7e7a8580-9b4c-11ea-84cc-2c3267d5a783.png)

And how I look at our current approach:

![image](https://user-images.githubusercontent.com/1880412/82568239-86d2c080-9b4c-11ea-8054-d28f4bc8c41a.png)

(Though of course there are always exceptions.)

All that to say, I recognize its complexity.

## Accession number from XML

That all seems reasonable. 

|  That could potentially be an interesting tack to consider taking with the AIPs as well if we were to rewrite the indexing functions? 

If the suggestion is at one point, these two sets of scripts would ideally be one that: 1. Consume the METS, 2. Index the METS. Then there's definitely a draw to that in terms of maintenance. 

An aside. I don't think I had  formed an opinion when reviewing about using [`metsrw`](https://adr.archivematica.org/0001-unification-of-mets-creation.html). Other than it maybe seemed out of scope. But the reindex script is already using it, and seems to make accession id easily available. Is it something to consider in scope of this work to use METS Reader-writer for any XML functions you're adopting? 

It looks like a lot of the XML work is already pre-existing, so perhaps it really is just out of scope for now.

Reply:
_Re: git hygeine_: I see your point here and the the diagrams make the difference between what's expected and what I submitted for this PR very clear. I've rebased on https://github.com/artefactual/archivematica/commit/ef6931b7a789b6009867bb607c58495cde246e98 and will be more mindful moving forward of my git practice. Genuinely, thank you for helping me with some of these growing pains.

_Re: XML_: 

> If the suggestion is at one point, these two sets of scripts would ideally be one that: 1. Consume the METS, 2. Index the METS. Then there's definitely a draw to that in terms of maintenance.

I wasn't thinking that so much as musing that the re-indexing script for AIPs not associated with a pipeline could borrow the approach from re-indexing transfers not associated with a pipeline (namely, reading information from the XML into the database, which enables the general indexing functions to be more efficient by retrieving information from the database rather an XML). Really just musing here, as 1) I don't think this is a priority and 2) that's definitely well out of scope for this work.

> An aside. I don't think I had formed an opinion when reviewing about using metsrw. Other than it maybe seemed out of scope. But the reindex script is already using it, and seems to make accession id easily available. Is it something to consider in scope of this work to use METS Reader-writer for any XML functions you're adopting?
> It looks like a lot of the XML work is already pre-existing, so perhaps it really is just out of scope for now.

My inclination is that using efficiently metsrw in the the parts of the AIP indexing functions I've modified for this work would more or less require rewriting the functions, and that seemed out of scope for now to me as well. It would be a great improvement though, I agree. Do you think logging an issue for it would be an okay compromise for the moment?

Reply:
NP @tw4l. If it's helpful at all that's good. These are small things in the context of much more significant work you're doing here :slightly_smiling_face: 

| It would be a great improvement though, I agree. Do you think logging an issue for it would be an okay compromise for the moment? 

That's okay. I don't think it needs a follow up issue. The general principle of replacing `metsrw` is a question we should ask in most PRs touching METS as a by-product of the ADR which I think is largely how we'll see it addressed in future. It would be great as its own maintenance task at some point. I think it's absolutely valid to be out of scope. And these are good reasons. It's not an insignificant undertaking.